File: Graphs/1135. Connecting Cities With Minimum Cost.cpp
class Solution {
public:
    static bool com(vector<int>& a, vector<int>& b) {
        return a[2] < b[2];
    }
    int find(int node, vector<int>& parents) {
        if (parents[node] != node) parents[node] = find(parents[node], parents);


        return parents[node];
    }

    int uf(int n1, int n2, int c, vector<int>& parents) {
        int p1 = find(n1, parents);
        int p2 = find(n2, parents);
        if (p1 == p2) return 0;
        parents[p1] = p2;
        return c;

    }


    int minimumCost(int n, vector<vector<int>>& connections) {
        sort(connections.begin(), connections.end(), com);
        vector<int> parents(n, 0);
        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }
        int res = 0;
        for (auto connect: connections) {
            res += uf(connect[0] - 1, connect[1] - 1, connect[2], parents);
        }
        unordered_set<int> visited;
        for (int par: parents) {
            visited.insert(find(par, parents));
        }
        if (visited.size() > 1) {
            return -1;
        }
        return res;
    }
};

File: Graphs/1135. Connecting Cities With Minimum Cost.py
class Solution:
    def minimumCost(self, n: int, connections: List[List[int]]) -> int:
        # if it is impossible, then return -1
        # we can use kruskal's algorithm
        connections.sort(key=lambda x: x[2])
        parents = [i for i in range(n)]

        def find(node):
            if parents[node] != node:
                parents[node] = find(parents[node])
            return parents[node]

        def uf(n1, n2, cost):
            p1 = find(n1)
            p2 = find(n2)
            if p1 == p2:
                return 0

            parents[p1] = p2
            return cost

        res = 0
        for a, b, cost in connections:
            res += uf(a - 1, b - 1, cost)
        visited = set()
        for par in parents:
            visited.add(find(par))

        if len(visited) != 1:
            return -1
        return res